### PR TITLE
Casst/documentation

### PR DESCRIPTION
**Pull request by casstait**
# Background

Added documentation in a ReadMe on how to install plugin and run and build on a local development machine

# How to review this PR
If you were to read this ReadMe for the first time would you be able to:
1. Install the plugin onto a TeamCity instance
2. Get your local development for plugin updates up and running

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,76 @@
 # TeamCity Build Opentelemetry Exporter Plugin
 
-A custom TeamCity plugin that contains a build listener to send OpenTelemetry trace data to an external endpoint
+A custom TeamCity plugin that registers a build listener to send build trace data to an opentelemetry collector external endpoint.
```
Comment (matt-richardson) [inline README.md:3]: This feels a bit too implementation specific 
```suggestion
A custom TeamCity plugin that sends build trace data to an OpenTelemetry collector endpoint.
```

File: README.md [hunk 2]
```diff
@@ -1,4 +1,76 @@
...
+You can use OpenTelemetry to instrument, generate, collect, and export telemetry data (metrics, logs, and traces) for analysis in order to understand your software's performance and behavior. Create and collect telemetry data from your services and software, then forward them to a variety of analysis tools. For more information about OpenTelemetry go to https://opentelemetry.io/. Documentation for instrumentation for Java and other languages can be found [here](https://opentelemetry.io/docs/).
+
+## How to use this plugin
+
+### Installing the plugin to TeamCity
+
+1. Build the plugin using the "Building" instructions below.
+2. In your TeamCity instance go to Administration -> Diagnostics -> Internal Properties.
+    1. Ensure you update add a property `octopus.teamcity.opentelemetry.plugin.endpoint=<your_opentelemetry_collector_endpoint>`
+    2. Ensure you update add a property `octopus.teamcity.opentelemetry.plugin.headers=<your_opentelemetry_collector_endpoint_required_headers>`
```
Comment (matt-richardson) [inline README.md:16]: 1. not sure we need `required` here.
2. add some info about separators
```suggestion
    2. Ensure you update add a property `octopus.teamcity.opentelemetry.plugin.headers=<your_opentelemetry_collector_endpoint_headers>`. Separate each key value pair with `=`, and separate each header with a `,`
```

File: README.md [hunk 3]
```diff
@@ -1,4 +1,76 @@
...
+3. Alternatively open the `internal.properties` file location in your TeamCity instance `data_dir/config` folder and update the above properties
+4. Copy to .zip to your TeamCity data_dir/plugins directory and restart TeamCity server OR install the .zip using your TeamCity instance UI
```
Comment (matt-richardson) [inline README.md:18]:
```suggestion
4. Install the .zip using your TeamCity instance UI via Administration -> Plugins -> Upload. Restart if required.
```

File: README.md [hunk 4]
```diff
@@ -1,4 +1,76 @@
...
+4. In a separate terminal, to run the default agent, cd in your TeamCity directory then run `./buildAgent/bin/agent.sh run`.
+5. Open localhost:8111 in a browser.
+6. On your first run create an admin login (this setup only needs to take place once due to the configuration mount). Once the server starts up, navigate to Agents->Unauthorized and authorise the agent that was started in a container alongside the server.
+7. (Optional) If attaching a remote debugger run in your TeamCity directory `export TEAMCITY_SERVER_OPTS=-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=8111 && $SCRIPT_PATH/bin/teamcity-server.sh run` for the server and `export TEAMCITY_SERVER_OPTS=-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=8111 && $SCRIPT_PATH/buildAgent/bin/agent.sh run` for the default agent.
+8. In your TeamCity instance go to Administration -> Diagnostics -> Internal Properties.
+   1. Ensure you update add a property `octopus.teamcity.opentelemetry.plugin.endpoint=<your_opentelemetry_collector_endpoint>`
+   2. Ensure you update add a property `octopus.teamcity.opentelemetry.plugin.headers=<your_opentelemetry_collector_endpoint_required_headers>`
```
Comment (matt-richardson) [inline README.md:36]: include the same separator info as above

File: README.md [hunk 5]
```diff
@@ -1,4 +1,76 @@
...
+3. Inside the root project folder run `./gradlew build`. The gradlew script will download Gradle for you if it is not already installed.
+4. The plugin is available at <project_root>/build/distributions/Octopus.TeamCity.OpenTelemetry.<version>.zip
+5. Copy to .zip to your TeamCity data_dir/plugins directory and restart TeamCity server OR install the .zip using your TeamCity instance UI
+
+### Testing
+
+JUnit test's have been added to package test folders.
+
+To test the plugin from code:
+1. Ensure your JAVA_HOME points to a java11 JDK installation
```
Comment (matt-richardson) [inline README.md:69]:
```suggestion
1. Ensure your `JAVA_HOME` points to a java11 JDK installation
```

File: README.md [hunk 6]
```diff
@@ -1,4 +1,76 @@
...
+2. Inside the root project folder run `./gradlew test`. The gradlew script will download Gradle for you if it is not already installed.
+
+### Cleaning
+
+To clean the project root directory of builds:
+3. Inside the root project folder run `./gradlew clean`. The gradlew script will download Gradle for you if it is not already installed.
```
Comment (matt-richardson) [inline README.md:75]:
```suggestion
1. Inside the root project folder run `./gradlew clean`. The gradlew script will download Gradle for you if it is not already installed.
```

File: README.md [hunk 7]
```diff
@@ -1,4 +1,76 @@
 # TeamCity Build Opentelemetry Exporter Plugin
 
-A custom TeamCity plugin that contains a build listener to send OpenTelemetry trace data to an external endpoint
+A custom TeamCity plugin that registers a build listener to send build trace data to an opentelemetry collector external endpoint.
+
+## What is OpenTelemetry
+
+You can use OpenTelemetry to instrument, generate, collect, and export telemetry data (metrics, logs, and traces) for analysis in order to understand your software's performance and behavior. Create and collect telemetry data from your services and software, then forward them to a variety of analysis tools. For more information about OpenTelemetry go to https://opentelemetry.io/. Documentation for instrumentation for Java and other languages can be found [here](https://opentelemetry.io/docs/).
+
```
Comment (matt-richardson) [inline README.md:8]: Add another section in here about what this gives us for builds (ie, why should i install it, what can I do with it). Maybe a picture too?

File: README.md [hunk 8]
```diff
@@ -1,4 +1,76 @@
...
+## How to use this plugin
+
```
Comment (matt-richardson) [inline README.md:10]: Might need some words "install it to teamcity, stuff will appear magically in your honeycomb/zipkin/jaeger installation"

File: README.md [hunk 9]
```diff
@@ -1,4 +1,76 @@
...
+### Installing the plugin to TeamCity
+
+1. Build the plugin using the "Building" instructions below.
+2. In your TeamCity instance go to Administration -> Diagnostics -> Internal Properties.
+    1. Ensure you update add a property `octopus.teamcity.opentelemetry.plugin.endpoint=<your_opentelemetry_collector_endpoint>`
+    2. Ensure you update add a property `octopus.teamcity.opentelemetry.plugin.headers=<your_opentelemetry_collector_endpoint_required_headers>`
+3. Alternatively open the `internal.properties` file location in your TeamCity instance `data_dir/config` folder and update the above properties
+4. Copy to .zip to your TeamCity data_dir/plugins directory and restart TeamCity server OR install the .zip using your TeamCity instance UI
+
+### Local Development
```
Comment (matt-richardson) [inline README.md:20]: I think the nesting here is 1 too deep

File: README.md [hunk 10]
```diff
@@ -1,4 +1,76 @@
...
+6. On your first run create an admin login (this setup only needs to take place once due to the configuration mount). Once the server starts up, navigate to Agents->Unauthorized and authorise the agent that was started in a container alongside the server.
+7. (Optional) If attaching a remote debugger run in your TeamCity directory `export TEAMCITY_SERVER_OPTS=-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=8111 && $SCRIPT_PATH/bin/teamcity-server.sh run` for the server and `export TEAMCITY_SERVER_OPTS=-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=8111 && $SCRIPT_PATH/buildAgent/bin/agent.sh run` for the default agent.
+8. In your TeamCity instance go to Administration -> Diagnostics -> Internal Properties.
+   1. Ensure you update add a property `octopus.teamcity.opentelemetry.plugin.endpoint=<your_opentelemetry_collector_endpoint>`
+   2. Ensure you update add a property `octopus.teamcity.opentelemetry.plugin.headers=<your_opentelemetry_collector_endpoint_required_headers>`
+9. To stop the TeamCity server and agent from running, in a separate terminal cd to your TeamCity directory and run `./bin/runAll.sh stop`
+
+##### Using Docker
```
Comment (matt-richardson) [inline README.md:39]: Recommend moving this up to promote a "docker first" approach

File: README.md [hunk 11]
```diff
@@ -1,4 +1,76 @@
...
+5. Copy to .zip to your TeamCity data_dir/plugins directory and restart TeamCity server OR install the .zip using your TeamCity instance UI
+
+### Testing
+
+JUnit test's have been added to package test folders.
+
+To test the plugin from code:
+1. Ensure your JAVA_HOME points to a java11 JDK installation
+2. Inside the root project folder run `./gradlew test`. The gradlew script will download Gradle for you if it is not already installed.
+
```
Comment (matt-richardson) [inline README.md:71]: Consider another section somewhere here about "how it gets built/deployed" (but be careful not to share too much detail about internal systems)

**Review by matt-richardson — CHANGES_REQUESTED** [review]
Nice one - this feels like a good step forward.
I'm not sure it tells me why I should use it though. I think we could get some inspiration from the jenkins plugin as to what it sends etc

File: README.md [hunk 12]
```diff
@@ -1,4 +1,76 @@
...
+## What is OpenTelemetry
+
+You can use OpenTelemetry to instrument, generate, collect, and export telemetry data (metrics, logs, and traces) for analysis in order to understand your software's performance and behavior. Create and collect telemetry data from your services and software, then forward them to a variety of analysis tools. For more information about OpenTelemetry go to https://opentelemetry.io/. Documentation for instrumentation for Java and other languages can be found [here](https://opentelemetry.io/docs/).
+
+## How to use this plugin
+
+### Installing the plugin to TeamCity
+
+1. Build the plugin using the "Building" instructions below.
```
[the remark this answers is not in the document]
Comment (casstait) [reply README.md:16]: We can update this as our pipeline changes, sure thing!

File: README.md (hunk 8 above)
Comment (casstait) [reply README.md:13]: added: 
"If you would like to visualize how you can better optimize your TeamCity builds and their dependency trees, this plugin exports TeamCity build pipeline data automatically to existing OpenTelemetry collector (such as honeycomb/zipkin/jaeger)."

File: README.md (hunk 7 above)
Comment (casstait) [reply README.md:11]: Added: 
"Example: A TeamCity composite build with various build dependancies and associated build steps with OpenTelemetry data represented in Honeycomb Traces." (image below)

File: README.md (hunk 10 above)
Comment (casstait) [reply README.md:42]: moved up!

File: README.md (hunk 11 above)
Comment (casstait) [reply README.md:74]: Currently the "Building" explains how the .zip gets built and the "Installing the plugin to TeamCity" or "Local Development" explains how to deploy it. We currently don't have any automated deployment.

**Review by matt-richardson — APPROVED** [review]
We paired through it, so :shipit:

**Pull request merged**